### ISSUE TITLE
dynamic_RNN fails on mac with unspecified batch_Size

### ISSUE THREAD

**Issue by pfaucon** [tensorflow/tensorflow#5223]
tl;dr:
dynamic_rnn fails on OS X with unspecified batch_size but works fine on Ubuntu 14.04/Python3.4.3.  It seems that on OS X the 'None != "?"' evaluates poorly, then at some point in ops/rnn.py +918 this equality fails:
 if const_batch_size != got_batch_size:
(i.e. None != Dimension("?"))

My best guess is that Dimension("?") is hacked in such a way that any comparison using it returns some fake None which for some reason on OSX + python 3.4 means true (I tried explicitly "if None:" as a sanity check, it evaluates as false).  For now I've modified the offending code to be: 
if const_batch_size != got_batch_size:
      if  (const_batch_size != got_batch_size) == got_batch_size or (const_batch_size != got_batch_size) == const_batch_size:
            print("weird dimensions stuff in ops/rnn.py, what the shit python?")
            continue
      raise ValueError(
          "Batch_size is not the same for all the elements in the input.")

This
### Environment info
Operating System:
OS X 10.12
Python 3.4.1
Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
N/A (CPU version)

If installed from binary pip package, provide:
0.11.0rc1 (just installed wheel)
also fails on 0.10.0rc1, 0.10.0 (though it works on 0.10.0 on Linux)

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code).  It may work with less but this was a simpl

import tensorflow as tf

input_size = 256
length = 100
num_classes = 64

data = tf.placeholder(tf.float32, [None, length, input_size])
target = tf.placeholder(tf.float32, [None, length, num_classes])

max_length = int(target.get_shape()[1])
num_classes = int(target.get_shape()[2])

network = tf.nn.rnn_cell.GRUCell(200)
output, new_state = tf.nn.dynamic_rnn(network, data, dtype=tf.float32)

### What other attempted solutions have you tried?
Works fine on my Linux box with or without GPU

### Logs or other output that would be helpful
(If logs are large, please upload as attachment or provide link).
Traceback (most recent call last):
  File "<my_directory>/tensorflow_mac_fail_test.py", line 18, in <module>
    output, new_state = tf.nn.dynamic_rnn(network, data, dtype=tf.float32)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/tensorflow/python/ops/rnn.py", line 836, in dynamic_rnn
    dtype=dtype)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/tensorflow/python/ops/rnn.py", line 920, in _dynamic_rnn_loop
    "Batch_size is not the same for all the elements in the input.")
ValueError: Batch_size is not the same for all the elements in the input.

**Comment by drpngx** [comment]
It looks like you pretty much know what the problem is. Do you have a idea for a fix you could send?

**Comment by pfaucon** [comment]
we could replace the existing comparison with:
if const_batch_size != got_batch_size and not (const_batch_size == got_batch_size) == None:

(if the dimensions aren't equal but also neither are unspecified) but it seems quite awkward.  I feel like a cleaner solution might be ensuring that None != Dimension("?") evaluates to false instead of 
None ==Dimension ("?")  => None, and the negation becomes true.

**Comment by drpngx** [comment]
@vrv possibly for any additional insight

**Comment by drpngx** [comment]
The fix seems reasonable to me, perhaps send a PR if you have one.

**Comment by pfaucon** [comment]
created patch https://github.com/tensorflow/tensorflow/compare/master...pfaucon:patch-1

**Comment by drpngx** [comment]
Thanks! Feel free to send the PR.
